extensions/security/deployment/src/main/java/io/quarkus/security/deployment/SecurityProcessor.java

```diff
@@ -203,61 +203,106 @@ private ResultHandle paramTypes(MethodCreator ctor, List<Type> parameters) {
-                AnnotationInstance methodAnnotation = getSingle(annotationStore.getAnnotations(method), securityAnnotations);
-                methodAnnotation = methodAnnotation == null ? classLevelAnnotation : methodAnnotation;
-                if (methodAnnotation != null) {
-                    methodAnnotations.put(method, methodAnnotation);
+        /*
+         * If we need to add the denyAll security check to all unannotated methods, we simply go through all secured methods,
+         * collect the declaring classes, then go through all methods of the classes and add the necessary check
+         */
+        if (denyUnannotated) {
+            Set<ClassInfo> allClassesWithSecurityChecks = new HashSet<>(methodToInstanceCollector.keySet().size());
+            for (MethodInfo methodInfo : methodToInstanceCollector.keySet()) {
+                allClassesWithSecurityChecks.add(methodInfo.declaringClass());
+            }
+            for (ClassInfo classWithSecurityCheck : allClassesWithSecurityChecks) {
+                for (MethodInfo methodInfo : classWithSecurityCheck.methods()) {
+                    if (!isPublicNonStaticNonConstructor(methodInfo)) {
+                        continue;
+                    }
+                    if (methodToInstanceCollector.containsKey(methodInfo)) { // the method already has a security check
+                        continue;
+                    }
+                    result.put(methodInfo, denyAllSecurityCheck());
                 }
             }
         }
-        return methodAnnotations;
+
+        return result;
     }
 
-    private AnnotationInstance getSingle(Collection<AnnotationInstance> classAnnotations, Set<DotName> securityAnnotations) {
-        AnnotationInstance result = null;
-        for (AnnotationInstance annotation : classAnnotations) {
-            if (securityAnnotations.contains(annotation.name())) {
-                if (result != null) {
-                    throw new IllegalStateException("Multiple security annotations on target: " + annotation.target());
+    private boolean isPublicNonStaticNonConstructor(MethodInfo methodInfo) {
+        return Modifier.isPublic(methodInfo.flags()) && !Modifier.isStatic(methodInfo.flags())
+                && !"<init>".equals(methodInfo.name());
+    }
+
+    private Map<MethodInfo, Function<BytecodeCreator, ResultHandle>> gatherSecurityAnnotations(
+            ApplicationIndexBuildItem indexBuildItem, DotName dotName,
+            Map<MethodInfo, AnnotationInstance> alreadyCheckedMethods,
+            Function<AnnotationInstance, Function<BytecodeCreator, ResultHandle>> securityCheckInstanceCreator) {
+
+        Map<MethodInfo, Function<BytecodeCreator, ResultHandle>> result = new HashMap<>();
+
+        List<AnnotationInstance> instances = indexBuildItem.getIndex().getAnnotations(dotName);
+        // make sure we process annotations on methods first
+        for (AnnotationInstance instance : instances) {
```

Comment:
It would be nice to have some more straightforward logic, without the need of `alreadyCheckedMethods` (you need them to, e.g., not override method level `@RolesAllowed` by class level `@PermitAll`, right?)

Maybe I'm missing something but I don't see the need to modify what was here before.

Reply:
oh, the cause for the change is probably that what I had before relied on scanning and you started coding it in a different way :)

Reply:
Basically I needed a way to both collect the annotations and at the same time determine the security check of the method.
So the logic needed to change from what you had